woocommerce/woocommerce-ios
# Login: Jetpack setup error handling

Part of #8075 


### Description

This PR adds error handling for the native Jetpack setup flow. There are 3 cases handled here:

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**
When the user taps the "Log in with another account" link on the Jetpack connection web page, they can log in with a different account other than the logged-in one. I handled this edge case by:
- Updated `SiteCredentialLoginViewModel` to return the site XMLRPC URL when the login completes.
- Updated the `LoginJetpackSetupViewModel` to return the Jetpack-connected email when the setup completes.
- Updated the coordinator to navigate to WP.com account verification if the Jetpack-connected email is different from the logged-in email.

**2. Error handling for setup steps.**
- Updated `LoginJetpackSetupViewModel` with information about the setup failure, error messages and codes, and visibility for error state views.
- `checkJetpackConnection` is handled with an auto-retry mechanism that allows a maximum of 2 retries. This is to ensure that Jetpack is synced properly for the site and returns the correct connection status.
- Updated `LoginJetpackSetupView` to show appropriate contents for error states.

**3. Handling interrupted Jetpack connection.**
When the user dismisses the web view for the Jetpack connection, we should show a modal indicating that the connection step has been interrupted. A new SwiftUI view has been added for this: `LoginJetpackSetupInterruptedView`.

Due to the size of the PR, unit tests will be added in a separate PR.

### Testing instructions

Prerequisite: make sure that you have a test site without Jetpack.

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for your store.
- On the Jetpack setup screen, when the setup reaches the connection step, notice that a web view is presented for approving Jetpack connection.
- Tap the "Log in with another account" on the web page and proceed to log in with a WP.com other than your logged-in account.
- Tap on the Authorize button on the web page, notice that the web view is then dismissed, and the Jetpack setup screens should then show that all steps are completed.
- Tap on Go to Store, notice that you are navigated to a new screen for verifying the password for the account you just entered on the web page. 
- After the login succeeds, notice that you are navigated to the home screen.

**2. Error handling for an account without permission.**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for an account **whose role is neither shop manager or administrator.**
- On the Jetpack setup screen, notice that the error state is then be displayed with a message regarding a permission issue. The retry CTA should not be available for this error. The Cancel and Get support CTAs should work properly.

**3. Error handling for issues other than permission**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for a shop manager or administrator account.
- When the Jetpack setup screen appears, turn off the internet connection or intercept the response with a 5xx error. Notice that the error state will then be displayed with appropriate messages. The retry CTA should be available and work properly.

**4. Error handling for interrupted Jetpack connection**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for a shop manager or administrator account.
- On the Jetpack setup screen, when the setup reaches the connection step, notice that a web view is presented for approving the Jetpack connection.
- Tap Cancel to dismiss the web view, notice that a new view is presented regarding the interrupted Jetpack connection.
- Tap Continue Connection, notice that the error screen is dismissed, and the web view is displayed again.
- Repeat tapping Cancel, this time tap Help, notice that the support screen is displayed.
- Tap the Cancel Installation button, notice that the installation flow is dismissed.

### Screenshots

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**

https://user-images.githubusercontent.com/5533851/203531637-52ce0fc7-63da-429e-ba7b-3989bd5ca038.mp4

**2. Error handling for an account without permission.**

https://user-images.githubusercontent.com/5533851/203531829-3b708d4a-9c1f-416a-8010-da30d76f55e8.mp4


**3. Error handling for issues other than permission**

https://user-images.githubusercontent.com/5533851/203531940-2e512b8a-e6be-41a8-8786-e3980c134a94.mp4

**4. Error handling for interrupted Jetpack connection**

https://user-images.githubusercontent.com/5533851/203532077-50ccd77e-5707-4b31-a90e-efb19e00cd96.mp4



---
